<a href="https://colab.research.google.com/github/Niraj7667/ML-models/blob/main/NaiveModels.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
from sklearn.datasets import load_wine
wine = load_wine()

In [4]:
dir(wine)

['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']

In [6]:
df = pd.DataFrame(wine.data, columns=wine.feature_names)
display(df.head())

alcohol  malic_acid   ash  ...   hue  od280/od315_of_diluted_wines  proline
0    14.23        1.71  2.43  ...  1.04                          3.92   1065.0
1    13.20        1.78  2.14  ...  1.05                          3.40   1050.0
2    13.16        2.36  2.67  ...  1.03                          3.17   1185.0
3    14.37        1.95  2.50  ...  0.86                          3.45   1480.0
4    13.24        2.59  2.87  ...  1.04                          2.93    735.0

[5 rows x 13 columns]

In [10]:
y = pd.DataFrame(wine.target, columns=['target'])
y

target
0         0
1         0
2         0
3         0
4         0
..      ...
173       2
174       2
175       2
176       2
177       2

[178 rows x 1 columns]

In [9]:
df['target'] = wine.target
display(df.head())

alcohol  malic_acid   ash  ...  od280/od315_of_diluted_wines  proline  target
0    14.23        1.71  2.43  ...                          3.92   1065.0       0
1    13.20        1.78  2.14  ...                          3.40   1050.0       0
2    13.16        2.36  2.67  ...                          3.17   1185.0       0
3    14.37        1.95  2.50  ...                          3.45   1480.0       0
4    13.24        2.59  2.87  ...                          2.93    735.0       0

[5 rows x 14 columns]

# Task
Split the dataframe `df` into 3 categories, then train Gaussian and Multinomial Naive Bayes classifiers on the training data, evaluate their performance on the test data, and determine which model performs better.

## Split data

### Subtask:
Split the DataFrame `df` into training and testing sets.


**Reasoning**:
Separate the features and target, then split the data into training and testing sets using `train_test_split`.



In [11]:
from sklearn.model_selection import train_test_split

X = df.drop('target', axis=1)
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Train gaussian naive bayes

### Subtask:
Train a Gaussian Naive Bayes classifier on the training data.


**Reasoning**:
Import the necessary classifier and train the Gaussian Naive Bayes model on the training data.



In [12]:
from sklearn.naive_bayes import GaussianNB

gnb = GaussianNB()
gnb.fit(X_train, y_train)

GaussianNB()

## Evaluate gaussian naive bayes

### Subtask:
Evaluate the performance of the Gaussian Naive Bayes classifier on the test data and display the results.


**Reasoning**:
Evaluate the performance of the trained Gaussian Naive Bayes classifier on the test data using various classification metrics and display the results.



In [17]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

y_pred_gnb = gnb.predict(X_test)

accuracy_gnb = accuracy_score(y_test, y_pred_gnb)
precision_gnb = precision_score(y_test, y_pred_gnb, average='weighted')
recall_gnb = recall_score(y_test, y_pred_gnb, average='weighted')
f1_gnb = f1_score(y_test, y_pred_gnb, average='weighted')


print("Gaussian Naive Bayes Performance:")
print(f"Accuracy: {accuracy_gnb:.4f}")
print(f"Precision: {precision_gnb:.4f}")
print(f"Recall: {recall_gnb:.4f}")
print(f"F1-score: {f1_gnb:.4f}")
print(f"model-score: {gnb.score(X_test,y_test):.4f}")

Gaussian Naive Bayes Performance:
Accuracy: 1.0000
Precision: 1.0000
Recall: 1.0000
F1-score: 1.0000
model-score: 1.0000


## Train multinomial naive bayes

### Subtask:
Train a Multinomial Naive Bayes classifier on the training data.


**Reasoning**:
Import the MultinomialNB class, instantiate it, and train the model using the training data.



In [14]:
from sklearn.naive_bayes import MultinomialNB

mnb = MultinomialNB()
mnb.fit(X_train, y_train)

MultinomialNB()

## Evaluate multinomial naive bayes

### Subtask:
Evaluate the performance of the Multinomial Naive Bayes classifier on the test data and display the results.


**Reasoning**:
Evaluate the performance of the trained Multinomial Naive Bayes model on the test data and display the calculated metrics.



In [20]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

y_pred_mnb = mnb.predict(X_test)

accuracy_mnb = accuracy_score(y_test, y_pred_mnb)
precision_mnb = precision_score(y_test, y_pred_mnb, average='weighted')
recall_mnb = recall_score(y_test, y_pred_mnb, average='weighted')
f1_mnb = f1_score(y_test, y_pred_mnb, average='weighted')

print("Multinomial Naive Bayes Performance:")
print(f"Accuracy: {accuracy_mnb:.4f}")
print(f"Precision: {precision_mnb:.4f}")
print(f"Recall: {recall_mnb:.4f}")
print(f"F1-score: {f1_mnb:.4f}")
print(f"model-score: {mnb.score(X_test,y_test):.4f}")

Multinomial Naive Bayes Performance:
Accuracy: 0.8889
Precision: 0.8866
Recall: 0.8889
F1-score: 0.8828
model-score: 0.8889


## Compare models

### Subtask:
Compare the performance of the two models and identify which one performs better.


**Reasoning**:
Compare the performance metrics of the two models and print the conclusion.



In [16]:
print("\nModel Comparison:")
if accuracy_gnb > accuracy_mnb:
    print("Gaussian Naive Bayes performed better than Multinomial Naive Bayes.")
    print(f"Gaussian Naive Bayes Accuracy: {accuracy_gnb:.4f}, Multinomial Naive Bayes Accuracy: {accuracy_mnb:.4f}")
elif accuracy_gnb < accuracy_mnb:
    print("Multinomial Naive Bayes performed better than Gaussian Naive Bayes.")
    print(f"Multinomial Naive Bayes Accuracy: {accuracy_mnb:.4f}, Gaussian Naive Bayes Accuracy: {accuracy_gnb:.4f}")
else:
    if f1_gnb > f1_mnb:
        print("Gaussian Naive Bayes performed better than Multinomial Naive Bayes.")
        print(f"Gaussian Naive Bayes F1-score: {f1_gnb:.4f}, Multinomial Naive Bayes F1-score: {f1_mnb:.4f}")
    elif f1_gnb < f1_mnb:
        print("Multinomial Naive Bayes performed better than Gaussian Naive Bayes.")
        print(f"Multinomial Naive Bayes F1-score: {f1_mnb:.4f}, Gaussian Naive Bayes F1-score: {f1_gnb:.4f}")
    else:
        print("Both models performed similarly based on the evaluated metrics.")



Model Comparison:
Gaussian Naive Bayes performed better than Multinomial Naive Bayes.
Gaussian Naive Bayes Accuracy: 1.0000, Multinomial Naive Bayes Accuracy: 0.8889


## Summary:

### Data Analysis Key Findings
* The Gaussian Naive Bayes classifier achieved perfect scores (1.0000) for accuracy, precision, recall, and F1-score on the test data.
* The Multinomial Naive Bayes classifier achieved an accuracy of 0.8889, precision of 0.8866, recall of 0.8889, and an F1-score of 0.8828 on the test data.
* Based on the evaluated metrics, the Gaussian Naive Bayes model performed better than the Multinomial Naive Bayes model.

### Insights or Next Steps
* The data might be linearly separable or have properties that are well-captured by the Gaussian distribution assumption in the Gaussian Naive Bayes model, leading to perfect classification.
* Investigate the features and their distributions to understand why Gaussian Naive Bayes performed perfectly, and consider if there are potential issues like data leakage or a very simple dataset that allows for perfect separation.
